### ISSUE TITLE
Stack overflow exception when concurrently creating two Dummies of type that takes a parameter of its own type 

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#1639]
https://ci.appveyor.com/project/blairconrad/fakeiteasy/builds/27840892#L279

I fear this is a symptom of a poorly-fixed #1603 (what else?), but figured I'd make the specific issue referencing the build failure until we can investigate.

**Comment by thomaslevesque** [comment]
Ouch, this is going to be hard to debug... I don't know if it's related to #1603; the fix looked pretty solid to me, but I don't remember all the details.

**Comment by blairconrad** [comment]
It is. Thanks! I was going for a solid fix. It's that we hadn't had stack overflows for a long long time (ever?), and my changes for that issue introduced a test that routinely overflowed its stack before the production code was fixed. So I made a wild guess.

**Comment by blairconrad** [comment]
I was able to make it happen again, and had `dotnet test` record the details about the tests run. I found that there were 3 tests that it started that it didn't gather results from:

* FakeItEasy.Specs.EventRaisingSpecs.DynamicRaiseAssignmentToNonDelegate
* FakeItEasy.Specs.NonGenericCreationSpecs.CollectionOfFakeWithOptionBuilder
* FakeItEasy.Specs.CallMatchingSpecs.ParameterHavingAnOutAttribute

I'll likely repeat this exercise to see if there's any commonality before diving in any deeper.

So I remember what I did later:

```powershell
1..300 | % { dotnet test -f netcoreapp1.0 -c release --no-build --nologo --blame --diag "out$_.log" }
```

And then ran

```powershell
.\analyze.py .\out12.host.19-10-03_06-53-29_98115_6.log   
```

where `analyze.py` is

```python
import sys

started = set()
finished = set()

f = open(sys.argv[1])
for line in f:
    if "RecordStart" in line:
        test_name = line.split(":")[6].strip()[:-1]
        started.add(test_name)
    elif "RecordResult" in line:
        test_name = line.split(":")[6].strip()[:-1]
        finished.add(test_name)

print(started.difference(finished))
```

**Comment by blairconrad** [comment]
Oh. The `--blame` probably didn't help. I should take that out.

**Comment by thomaslevesque** [comment]
Have you tried repeating the test run with only those tests?

**Comment by blairconrad** [comment]
I tried repeating with only those tests. In 300 runs, I had no failures. Retried with all tests. In 300 runs, I had 9 failures. Here were the tests with ungathered results:

* CancellationSpecs.AsyncWithoutResultCanceledTokenWithConfiguredCallForAnyToken EventRaisingSpecs.DynamicRaiseAssignmentToNonDelegate OrderedCallMatchingSpecs.OrderedAssertionsUsingRepeatedInOrder         
* CallMatchingSpecs.ParameterHavingAnOutAttribute CancellationSpecs.AsyncWithResultNonCanceledToken ThrowsAsyncSpecs.WithResultPassExceptionFactoryWithFourArgs 
* CallMatchingSpecs.FailingMatchOfNonGenericCalls ConfiguringPropertySetterSpecs.ConfiguringSetterToDoNothingForAnyValue FakingDelegates.Throws     
* ConfigurationSpecs.DoesNothingAfterStrictFakeableReferenceTypeKeepsDefaultReturnValue EventRaisingSpecs.DynamicRaiseAssignmentToNonDelegate                         
* AnyCallConfigurationSpecs.WithNonVoidReturnType AssertingCallCountWithRepeatedSpecs.CallCountConstraintWithNonMatchingNumberOfCalls CallMatchingSpecs.PassingAFakeWithABadToStringToAMethod EventRaisingSpecs.EventArgumentsThatDoNotExtendEventArg                      
* GenericCreationOptionsSpecs.CallsBaseMethodsOverridesConfigureFake LazySpecs.LazyReturnValueOfNonDummyableType ThenSpecs.AnyCallWithInvokesAndThrows UserCallbackExceptionSpecs.ExceptionInArgumentMatcher           
* CallsBaseMethodsFakeSpecs.AbstractMethod             
* EventRaisingSpecs.DynamicRaiseAssignmentToNonDelegate       
* ConfiguringPropertySetterSpecs.ConfiguringSetterFromBaseInterfaceOnFakedClass EventRaisingSpecs.DynamicRaiseAssignmentToNonDelegate UserCallbackExceptionSpecs.ExceptionInWhenArgumentsMatchPredicateForAnyCallSpec         

So. I need another tack!

**Comment by blairconrad** [comment]
Removed one instance of `AvoidLongSelfReferentialConstructor` spec (we have for both generic and non-generic), and ran for well over 300 iterations with no failure.
Restored it, and took a guess at a couple of places that might be issues (specifically using `[]` to add cached constructors and parameter types when making dummies and fakes respectively, and got this:

```
[xUnit.net 00:00:03.56]     FakeItEasy.Specs.NonGenericDummyCreationSpecs.AvoidLongSelfReferentialConstructor() [07] And it was created using the one-parameter constructor [FAIL]
```

I'm thinking this is a manifestation of would've been the stack overflow!
Now all I have to do is try to figure out what's happening down in the bowels.

**Comment by blairconrad** [comment]
I think I figured it out! Writing a test may be hard!

**Comment by thomaslevesque** [comment]
Nice work! So what's going on?

**Comment by blairconrad** [comment]
It's sort of the same as the original problem, only on 2 threads. Remember the class that has 2 constructors, one that takes a parameter of the its own type and another type, and one that just takes another type.

A1. We start making a Dummy of the class. There's no DummyFactory and nothing's in any of our caches.
A2. We eventually get down to making a Fake of the class. We try the 2-parameter constructor, which requires a Dummy of the type. The loop-detecting resolver figures out we're in a loop and that fails (before fixing #1603, we would've stack overflowed now).
A3. We fall back to the 1-parameter constructor which will eventually succeed
A4. Cache that the 1-parameter constructor is best
A5. Cache that the fake-creating resolver is our favourite resolver.
A7. Client gets a dummy. Yay!

Okay. Now thread B comes in 
B1. We start making a Dummy of the class. There's no DummyFactory and nothing's in any of our caches.
B2. We eventually get down to making a Fake of the class. We try the 2-parameter constructor, which requires a Dummy of the type.
B6. (After A5 - the timing of B1 and B2 didn't really matter that much.) Because there's a cached dummy resolver, we use it to create the Dummy that we need for the 2-parameter constructor. _We do not check the loop-detecting resolver in this case._
B7. The 2-parameter constructor succeeds! Yay.
B8. Cache that the 2-parameter constructor is best.
B9. We use TryAdd back in DummyValueResolver, so we don't overwrite the favourite resolver.
B10. The client gets their Dummy.

I believe that moving the loop-detecting resolver check above the check for the cached resolver in DummyValueResolver will fix this. I'll try to make a test that reliably reproduces it and then try the fix, though.

It's lucky that the generic and non-generic versions of our creation tests dummied the same class, or we never would've found this.

**Comment by blairconrad** [comment]
Oh! And if I don't update the caches to use `TryAdd`, sometimes we'll end up keeping the 2-parameter constructor as the preferred one, and that sets us up for the stack overflows!

**Comment by thomaslevesque** [comment]
Ah, I see... What a mess! I wonder if we shouldn't just introduce a lock (with double-checked locking: check the cache, take a lock if it's not in cache, check the cache again inside the lock). It would have a small performance impact, but it would probably make things simpler.

**Comment by blairconrad** [comment]
This would be around the `DummyValueResolver.strategyCache`? So we'd only ever have one thread try to figure out out how to make a Dummy? We'd have the first thread in (or unlikely, the first couple of threads) enter the lock?

```c#
public CreationResult TryResolveDummyValue(Type typeOfDummy, LoopDetectingResolutionContext resolutionContext)
{
    if (this.strategyCache.TryGetValue(typeOfDummy, out ResolveStrategy cachedStrategy))
    {
        return cachedStrategy.TryCreateDummyValue(typeOfDummy, this, resolutionContext);
    }

    lock (…)
    {
        if (this.strategyCache.TryGetValue(typeOfDummy, out ResolveStrategy cachedStrategy))
        {
            return cachedStrategy.TryCreateDummyValue(typeOfDummy, this, resolutionContext);
        }

        if (!resolutionContext.TryBeginToResolve(typeOfDummy))
        {
            return CreationResult.FailedToCreateDummy(typeOfDummy, "Recursive dependency detected. Already resolving " + typeOfDummy + '.');
        }

        var creationResult = this.TryResolveDummyValueWithAllAvailableStrategies(typeOfDummy, resolutionContext);
        if (creationResult.WasSuccessful)
        {
             resolutionContext.OnSuccessfulResolve(typeOfDummy);
        }
    }

    return creationResult;
}
```

That would probably fix it, with the advantage that we'd never do double work to figure out how to make a particular top-level Dummy, but we still would need all our safeguards in place at the other levels (protecting cached constructors chosen for activating objects, or constructor types in `FakeObjectCreator`, since we'd still be updating those caches from different threads.

**Comment by thomaslevesque** [comment]
Yes, exactly. The second thread to enter the lock will find the strategy in the cache and exit immediately.

> but we still would need all our safeguards in place at the other levels (protecting cached constructors chosen for activating objects, or constructor types in `FakeObjectCreator`, since we'd still be updating those caches from different threads.

Yes. Or we could also put a lock there...

**Comment by thomaslevesque** [comment]
This change has been released as part of [FakeItEasy 5.3.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/5.3.0).

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 5.3.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/5.3.0).
